qa/sc_evm_native_interop.py

```diff
@@ -157,6 +174,163 @@ def run_test(self):
+        assert_equal(storage_contract_address.lower(), evm_call["to"].lower())
+        assert_true(int(native_call["gas"], 16) > 0)
+        assert_true(int(native_call["gasUsed"], 16) > 0)
+        assert_equal(sol_contract_call_data_retrieve.lower(), evm_call["input"].lower())
+        output = int(evm_call["output"], 16)
+        assert_equal(initial_value, output)
+        assert_false("calls" in evm_call)
+
+        # Get gas estimations
+        estimation_interop = node.rpc_eth_estimateGas(
+            {
+                "to": native_contract_address,
+                "input": data_input
+            }
+        )
+        estimation_evm = node.rpc_eth_estimateGas(
+            {
+                "to": storage_contract_address,
+                "input": sol_contract_call_data_retrieve
+            }
+        )
+        logging.info("estimated gas interop: {}".format(estimation_interop))
+        logging.info("estimated gas evm: {}".format(estimation_evm))
+
+        # Gas usage given in a trace does not include intrinsic gas, we need to add it to compare with gas estimation
+        # 21k + (number of non-zero bytes in the input) * 16 + (number of zero bytes) * 4
+        intrinsic_gas = 21000 + 4 * 16
+
+        # Verify gas usage reported by the trace matches with the estimated gas for a call to the EVM contract
+        assert_equal(int(trace_result["gasUsed"], 16), int(estimation_interop["result"], 16))
+
+        # Verify gas usage of the nested call to the native contract reported by the trace matches with the estimation
+        assert_equal(int(evm_call["gasUsed"], 16) + intrinsic_gas, int(estimation_evm["result"], 16))
+
+
+
+
+        # Verify STATICCALL to a readwrite EVM contract method throws an error
+        method = 'invokeStaticCall(address,bytes)'
+        abi_str = function_signature_to_4byte_selector(method)
+        encoded_abi_method_signature = encode_hex(abi_str)
+        addr_padded_str = "000000000000000000000000" + remove_0x_prefix(storage_contract_address)
+        data_input_failed = encoded_abi_method_signature + addr_padded_str
+        data_input_failed += "0000000000000000000000000000000000000000000000000000000000000040"
+        h_len = hex(len(remove_0x_prefix(sol_contract_call_data_inc)) // 2)
+        data_input_failed += "000000000000000000000000000000000000000000000000000000000000000" + remove_0x_prefix(HexStr(h_len))
+        data_input_failed += remove_0x_prefix(sol_contract_call_data_inc)
+        data_input_failed += "00000000000000000000000000000000000000000000000000000000"
+
+        result = node.rpc_eth_call(
+            {
+                "to": native_contract_address,
+                "input": data_input_failed
+            }, "latest"
+        )
+
+        assert_true('error' in result)
+        assert_true('write protection' in result['error']['message'])
+
+        # TODO traceCall seems to have problems with failed txs. Needs to check what is the correct behavior
```

Comment:
Is this expected to be covered with this branch?

Reply:
No, there is a Jira ticket for this issue: https://horizenlabs.atlassian.net/browse/SDK-1434
It is solved in this PR:
https://github.com/HorizenOfficial/Sidechains-SDK/pull/924